app/assets/javascripts/vocab_sheet.js

```diff
@@ -91,51 +72,91 @@ $(document).ready(function() {
+      var maxLength = elem.attr('maxlength');
 
       if (notes.length >= maxLength) {
         elem.addClass('max-length-reached');
 
         setTimeout(function() {
-          textBox.removeClass('max-length-reached');
+          textBoxes.removeClass('max-length-reached');
         }, 1000);
       }
     }
 
     function setCharacterCount(elem) {
-      clearTimeout(typeTimer);
+      var notes = elem.val();
+      var maxLength = elem.attr('maxlength');
 
-      typeTimer = setTimeout(function() {
-        notes = elem.val();
+      elem.siblings('.character-count__wrap')
+        .children('.character-count__count')
+        .text(maxLength - notes.length);
 
-        elem.siblings('.character-count__wrap')
-          .children('.character-count__count')
-          .text(500 - notes.length);
-
-        updateNotes(formAction, signId, notes);
-      }, 100);
-    }
-
-    function updateNotes(action, signId, notes) {
-      $.ajax({
-        url: action,
-        method: 'PUT',
-        data: {
-          sign_id: signId,
-          notes: notes,
-        },
-        headers: {
-          'X-CSRF-Token': $('meta[name="authenticity-token"]').attr('content'),
-        },
-      }).done(function(data) {
-        // console.log("success!", data)
-      }).fail(function(error) {
-        console.error(error.statusText);
-      });
+      action = getFormAction(elem);
+      signId = getSignIdFromAction(action);
     }
 
     checkCharacterCount();
   };
 
+  if ($('.vocab-sheet__text-input').length > 0) {
+    $('.vocab-sheet__text-input').keyup(function() {
+      var field = $(this);
+      action = getFormAction(field);
+      signId = getSignIdFromAction(action);
+      var data = {signId};
+
+      if (field.hasClass('item-name')) data['name'] = field.val();
+      if (field.hasClass('item-maori-name')) data['maoriName'] = field.val();
+      if (field.hasClass('item-notes')) data['notes'] = field.val();
+
+      updateVocabItem(action, data);
+    });
+  }
+
+  function updateVocabItem(action, params) {
+    checkForTextTimeout(function(action, params) {
+      if (params.signId !== null) {
+        $.ajax({
+          url: action || '/vocab_sheet/items/' + params.signId,
+          method: 'PUT',
+          data: assignItemParams(params),
+          headers: {
+            'X-CSRF-Token': $('meta[name="authenticity-token"]').attr('content'),
+          },
+        }).fail(function(error) {
+          console.error(error.statusText);
+        });
+      } else {
+        console.error('Error, cannot update vocab sheet item without ID. Parameters presented:', params);
+      }
+    });
+  }
+
+  function assignItemParams(params) {
+    var data = {};
+
+    if (params.signId !== null) data['sign_id'] = params.signId;
+    if (params.name !== null) data['name'] = params.name;
+    if (params.maoriName !== null) data['maori_name'] = params.maoriName;
+    if (params.notes !== null) data['notes'] = params.notes;
```

Comment:
gosh this bit is repetitive. Is there a way to DRY it? :elephant:

Reply:
It is, right? The reason I did it this way was two-fold:

- The property names are different between Ruby and JS
- If the item isn't present in params, it shouldn't be in data otherwise it overrides it in Ruby (otherwise I would do a fallback assignment)

What do you think?